chalk/core.py

```diff
@@ -63,7 +69,14 @@ def _style(self, style: Style) -> Diagram:
     def compose(
         self, envelope: Envelope, other: Optional[Diagram] = None
     ) -> Diagram:
-        return Compose(envelope, self, other if other is not None else Empty())
+        if isinstance(self, Compose):
+            if isinstance(other, Compose) and other is not None:
+                return Compose(envelope, self.diagrams + other.diagrams)
+            if other is not None:
+                return Compose(envelope, list(self.diagrams) + [other])
+        return Compose(
+            envelope, [self, other if other is not None else Empty()]
+        )
```

Comment:
The logic here might be a bit difficult to follow (at least for me). For example, it seems that we are missing a case: when `self` is not `Compose`, but `other` is `Compose`. For this case, we should concatenate `self` wrapped as list with the diagrams unpacked from `other`.

What about something along these lines:
```python
other = other if other is not None else Empty()
if isinstance(self, Compose) and isinstance(other, Compose):
    return Compose(envelope, self.diagrams + other.diagrams)
elif isinstance(self, Compose):
    return Compose(envelope, self.diagrams + [other])
elif isinstance(other, Compose):
    return Compose(envelope, [self] + other.diagrams)
else:
    return Compose(envelope, [self, other])
```

Or with pattern matching (but this requires Python 3.10+):
```python
other = other if other is not None else Empty()
match (self, other):
    case (Compose(_, ds1), Compose(_, ds2)):
        return Compose(envelope, ds1 + ds2)
    case (Compose(_, ds1), _):
        return Compose(envelope, ds1 + [other])
    case (_, Compose(_, ds2)):
        return Compose(envelope, [self] + ds2)
    case (_, _):
        return Compose(envelope, [self, other])
```